Fix mova alignment issue in macro.

Signed-off-by: Carl Shaw <carl.shaw@st.com>

Index: linux-2.6.23/include/asm-sh/processor.h
===================================================================
--- linux-2.6.23.orig/include/asm-sh/processor.h	2008-04-17 13:20:46.000000000 +0100
+++ linux-2.6.23/include/asm-sh/processor.h	2008-04-17 13:21:26.000000000 +0100
@@ -20,7 +20,7 @@
  * Default implementation of macro that returns current
  * instruction pointer ("program counter").
  */
-#define current_text_addr() ({ void *pc; __asm__("mova	1f, %0\n1:":"=z" (pc)); pc; })
+#define current_text_addr() ({ void *pc; __asm__("mova	1f, %0\n.align 2\n1:":"=z" (pc)); pc; })
 
 /* Core Processor Version Register */
 #define CCN_PVR		0xff000030
