--- ZD1211LnxDrv_2_22_0_0.orig/src/zddebug.c	2007-11-09 17:16:17.000000000 +0100
+++ ZD1211LnxDrv_2_22_0_0/src/zddebug.c	2011-05-09 18:00:06.000000000 +0200
@@ -608,6 +608,13 @@
 {
 
 }
+
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,18))
+static inline void * __deprecated bus_to_virt(unsigned long address)
+{
+	return __va(address);
+}
+#endif
 
 int zd1205_zd_dbg_ioctl(struct zd1205_private *macp, struct zdap_ioctl *zdreq)
 {
