diff -Naur org/src/zddebug.c patched/src/zddebug.c
--- org/src/zddebug.c	2007-11-09 17:16:17.000000000 +0100
+++ patched/src/zddebug.c	2011-05-11 09:45:06.000000000 +0200
@@ -39,6 +39,12 @@
 extern U8 mMacMode;
 #if ZDCONF_APC == 1
 extern U8 APC_NT[256][6];
+#endif
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,18))
+static inline void * __deprecated bus_to_virt(unsigned long address)
+{
+	return __va(address);
+}
 #endif
 void zd1205_list_STAs(struct zd1205_private *macp)
 {

