Index: linux-2.6.23.1-stm/arch/sh/kernel/setup.c
===================================================================
--- linux-2.6.23.1-stm.orig/arch/sh/kernel/setup.c
+++ linux-2.6.23.1-stm/arch/sh/kernel/setup.c
@@ -112,6 +112,17 @@ static int __init early_parse_mem(char *
 	memory_start = (unsigned long)PAGE_OFFSET+__MEMORY_START;
 #endif
 	size = memparse(p, &p);
+
+	if (size > __MEMORY_SIZE) {
+		static char msg[] __initdata = KERN_ERR
+			"Using mem= to increase the size of kernel memory "
+			"is not allowed.\n"
+			"  Recompile the kernel with the correct value for "
+			"CONFIG_MEMORY_SIZE.\n";
+		printk(msg);
+		return 0;
+	}
+
 	memory_end = memory_start + size;
 
 	return 0;
