From ed18271e063fcaba6da3e92268ad5866c2ac8de9 Mon Sep 17 00:00:00 2001
From: Carl Shaw <carl.shaw@st.com>
Date: Thu, 19 Feb 2009 15:41:49 +0000
Subject: [PATCH] sh_stm: STMicroelectronics STLinux 2.3 kernel release 119

Signed-off-by: Carl Shaw <carl.shaw@st.com>
---
 localversion-stm |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/localversion-stm b/localversion-stm
index ae24526..f0d32ea 100644
--- a/localversion-stm
+++ b/localversion-stm
@@ -1 +1 @@
-_stm23_0118
+_stm23_0119
-- 
1.5.6.3

