Patch to correctly configure the OnChip SH4 RealTimeClock (RTC).

Signed-off-by: Angelo Castello <angelo.castello@st.com>
Index: linux/arch/sh/kernel/cpu/sh4/setup-stb7100.c
===================================================================
--- linux.orig/arch/sh/kernel/cpu/sh4/setup-stb7100.c	2008-01-23 12:46:09.000000000 +0000
+++ linux/arch/sh/kernel/cpu/sh4/setup-stb7100.c	2008-01-23 12:46:09.000000000 +0000
@@ -68,29 +68,6 @@
 	.resource = wdt_resource,
 };
 
-static struct resource rtc_resource[]= {
-	[0] = {
-		.start = 0xffc80000,
-		.end   = 0xffc80000 + 0x40,
-		.flags = IORESOURCE_MEM
-	},
-	[1] = {
-		.start = 20,/* Alarm IRQ   */
-		.flags = IORESOURCE_IRQ
-	},
-	[2] = {
-		.start = 21,/* Periodic IRQ*/
-		.flags = IORESOURCE_IRQ
-	},
-};
-
-static struct platform_device rtc_device = {
-	.name		= "rtc",
-	.id		= -1,
-	.num_resources	= ARRAY_SIZE(rtc_resource),
-	.resource	= rtc_resource,
-};
-
 /* USB resources ----------------------------------------------------------- */
 
 static struct resource st40_ohci_resources[] = {
@@ -628,6 +605,37 @@
 	platform_device_register(&stm_pwm_device);
 }
 
+/* SH-RTC resources ----------------------------------------------------------- */
+static struct resource rtc_resource[]= {
+        [0] = {
+		.start = 0xffc80000,
+		.end   = 0xffc80000 + 0x3c,
+	        .flags = IORESOURCE_IO
+	},
+	[1] = { /* periodic irq */
+		.start = 21,
+		.end   = 21,
+	        .flags = IORESOURCE_IRQ
+	},
+	[2] = { /* carry irq */
+		.start = 22,
+		.end   = 22,
+	        .flags = IORESOURCE_IRQ
+	},
+	[3] = { /* alarm irq */
+		.start = 20,
+		.end   = 20,
+	        .flags = IORESOURCE_IRQ
+	},
+};
+
+static struct platform_device rtc_device = {
+	.name           = "sh-rtc",
+	.id             = -1,
+	.num_resources  = ARRAY_SIZE(rtc_resource),
+	.resource       = rtc_resource,
+};
+
 /* LiRC resources ---------------------------------------------------------- */
 static struct lirc_pio lirc_pios[] = {
 	[0] = {
@@ -1124,10 +1132,10 @@
 static struct platform_device *stx710x_devices[] __initdata = {
 	&sci_device,
 	&wdt_device,
-	&rtc_device,
 	&fdma_710x_device,
 	&sysconf_device,
 	&ilc3_device,
+	&rtc_device,
 };
 
 static int __init stx710x_devices_setup(void)
Index: linux/arch/sh/kernel/cpu/sh4/setup-stx7200.c
===================================================================
--- linux.orig/arch/sh/kernel/cpu/sh4/setup-stx7200.c	2008-01-23 12:46:09.000000000 +0000
+++ linux/arch/sh/kernel/cpu/sh4/setup-stx7200.c	2008-01-23 12:46:09.000000000 +0000
@@ -689,6 +689,37 @@
 	platform_device_register(&stm_pwm_device);
 }
 
+/* SH-RTC resources ----------------------------------------------------------- */
+static struct resource rtc_resource[]= {
+        [0] = {
+		.start = 0xffc80000,
+		.end   = 0xffc80000 + 0x3c,
+	        .flags = IORESOURCE_IO
+	},
+	[1] = { /* periodic irq */
+		.start = 21,
+		.end   = 21,
+	        .flags = IORESOURCE_IRQ
+	},
+	[2] = { /* carry irq */
+		.start = 22,
+		.end   = 22,
+	        .flags = IORESOURCE_IRQ
+	},
+	[3] = { /* alarm irq */
+		.start = 20,
+		.end   = 20,
+	        .flags = IORESOURCE_IRQ
+	},
+};
+
+static struct platform_device rtc_device = {
+	.name           = "sh-rtc",
+	.id             = -1,
+	.num_resources  = ARRAY_SIZE(rtc_resource),
+	.resource       = rtc_resource,
+};
+
 /* LiRC resources ---------------------------------------------------------- */
 static struct lirc_pio lirc_pios[] = {
 	[0] = {
@@ -1365,6 +1396,12 @@
 	 * clock which is derived from the SATA clock. */
 	ctrl_outl(0, 0xFD701048);
 
+	/* Configure the ST40 RTC to source its clock from clockgenB.
+	 * In theory this should be board specific, but so far nobody
+	 * has ever done this. */
+	sc = sysconf_claim(SYS_CFG, 8, 1, 1, "rtc");
+	sysconf_write(sc, 1);
+
 	/* We haven't configured the LPC, so the sleep instruction may
 	 * do bad things. Thus we disable it here. */
 	disable_hlt();
@@ -1401,6 +1438,7 @@
 	&fdma_xbar_device,
 	&sysconf_device,
 	&ilc3_device,
+        &rtc_device,
 };
 
 static int __init stx7200_devices_setup(void)
