Index: linux-2.6.23.1-stm/arch/sh/mm/pmb.c
===================================================================
--- linux-2.6.23.1-stm.orig/arch/sh/mm/pmb.c
+++ linux-2.6.23.1-stm/arch/sh/mm/pmb.c
@@ -27,6 +27,7 @@
 #include <asm/pgtable.h>
 #include <asm/mmu.h>
 #include <asm/io.h>
+#include <asm/mmu_context.h>
 
 #define DPRINTK(fmt, args...) printk(KERN_ERR "%s: " fmt, __FUNCTION__, ## args)
 
@@ -430,6 +431,11 @@ apply_boot_mappings(struct pmb_mapping *
 		entry = entry->next;
 	} while (entry);
 
+	/* Flush out the TLB */
+	i =  ctrl_inl(MMUCR);
+	i |= MMUCR_TI;
+	ctrl_outl(i, MMUCR);
+
 	back_to_cached();
 }
 
Index: linux-2.6.23.1-stm/include/asm-sh/cpu-sh4/mmu_context.h
===================================================================
--- linux-2.6.23.1-stm.orig/include/asm-sh/cpu-sh4/mmu_context.h
+++ linux-2.6.23.1-stm/include/asm-sh/cpu-sh4/mmu_context.h
@@ -22,6 +22,8 @@
 #define MMU_UTLB_ADDRESS_ARRAY	0xF6000000
 #define MMU_PAGE_ASSOC_BIT	0x80
 
+#define MMUCR_TI		(1<<2)
+
 #ifdef CONFIG_X2TLB
 #define MMUCR_ME		(1 << 7)
 #else
